File: backend/app/storage.py
import hashlib
import logging
import os
import shutil
import uuid
from collections.abc import AsyncIterator
from datetime import datetime, timedelta, timezone
from pathlib import Path

import aiofiles
import jwt
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import settings

logger = logging.getLogger(__name__)


def storage_root() -> Path:
    return Path(settings.storage_path).resolve()


def files_dir() -> Path:
    return storage_root() / "files"


def temp_dir() -> Path:
    return storage_root() / "tmp"


def ensure_storage_dirs() -> None:
    files_dir().mkdir(parents=True, exist_ok=True)
    temp_dir().mkdir(parents=True, exist_ok=True)


def get_file_path(sha256: str) -> Path:
    sha = sha256.lower()
    return files_dir() / sha[:2] / sha[2:4] / sha


def get_temp_path(file_id: uuid.UUID | str) -> Path:
    return temp_dir() / str(file_id)


def file_exists(sha256: str) -> bool:
    return get_file_path(sha256).is_file()


def checksum_sha256(data: bytes) -> str:
    return hashlib.sha256(data).hexdigest().lower()


def create_download_ticket(item_id: uuid.UUID, file_id: uuid.UUID) -> tuple[str, datetime]:
    expires_at = datetime.now(timezone.utc) + timedelta(seconds=settings.upload_url_ttl_seconds)
    payload = {
        "sub": "download",
        "item_id": str(item_id),
        "file_id": str(file_id),
        "exp": int(expires_at.timestamp()),
    }
    ticket = jwt.encode(payload, settings.jwt_secret, algorithm="HS256")
    return ticket, expires_at


def verify_download_ticket(ticket: str, item_id: uuid.UUID, file_id: uuid.UUID) -> bool:
    try:
        payload = jwt.decode(ticket, settings.jwt_secret, algorithms=["HS256"])
        return (
            payload.get("sub") == "download"
            and payload.get("item_id") == str(item_id)
            and payload.get("file_id") == str(file_id)
        )
    except jwt.PyJWTError:
        return False


async def save_upload_stream(
    file_id: uuid.UUID | str,
    expected_sha256: str,
    expected_size: int,
    stream: AsyncIterator[bytes],
) -> Path:
    ensure_storage_dirs()
    temp_path = get_temp_path(file_id)
    hasher = hashlib.sha256()
    bytes_written = 0

    try:
        async with aiofiles.open(temp_path, "wb") as f:
            async for chunk in stream:
                if not chunk:
                    continue
                bytes_written += len(chunk)
                if bytes_written > settings.max_file_size:
                    raise ValueError("File size exceeds MAX_FILE_SIZE")
                hasher.update(chunk)
                await f.write(chunk)

        if bytes_written != expected_size:
            raise ValueError(f"Size mismatch: expected {expected_size}, got {bytes_written}")

        computed_sha = hasher.hexdigest().lower()
        if computed_sha != expected_sha256.lower():
            raise ValueError(f"SHA-256 mismatch: expected {expected_sha256.lower()}, got {computed_sha}")

        target_path = get_file_path(computed_sha)
        target_path.parent.mkdir(parents=True, exist_ok=True)
        # Atomic replace or shutil.move if across filesystems
        if temp_path.exists():
            shutil.move(str(temp_path), str(target_path))
        return target_path
    except Exception:
        if temp_path.exists():
            temp_path.unlink(missing_ok=True)
        raise


async def delete_file_if_unreferenced(sha256: str, session: AsyncSession) -> bool:
    from app.models import DropFile

    stmt = select(func.count()).select_from(DropFile).where(DropFile.sha256 == sha256)
    result = await session.execute(stmt)
    count = result.scalar_one()

    if count == 0:
        path = get_file_path(sha256)
        if path.is_file():
            path.unlink(missing_ok=True)
            logger.info("Deleted physical file %s", path)
            return True
    return False


def cleanup_temp_files(max_age_seconds: int = 3600) -> int:
    ensure_storage_dirs()
    cutoff = datetime.now(timezone.utc).timestamp() - max_age_seconds
    removed = 0
    for entry in temp_dir().iterdir():
        if entry.is_file():
            try:
                if entry.stat().st_mtime < cutoff:
                    entry.unlink(missing_ok=True)
                    removed += 1
            except Exception as e:
                logger.warning("Failed to remove temp file %s: %s", entry, e)
    return removed

File: backend/internal/storage/storage.go
package storage

import (
	"context"
	"crypto/sha256"
	"database/sql"
	"encoding/hex"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"github.com/google/uuid"
)

type StorageManager struct {
	storageRoot string
	jwtSecret   []byte
	ttlSeconds  int
	maxFileSize int64
}

type PhysicalFileInfo struct {
	SHA256   string
	Size     int64
	Path     string
	FullPath string
}

func NewStorageManager(storagePath string, jwtSecret string, ttlSeconds int, maxFileSize int64) *StorageManager {
	absPath, err := filepath.Abs(storagePath)
	if err != nil {
		absPath = storagePath
	}
	return &StorageManager{
		storageRoot: absPath,
		jwtSecret:   []byte(jwtSecret),
		ttlSeconds:  ttlSeconds,
		maxFileSize: maxFileSize,
	}
}

func (s *StorageManager) FilesDir() string {
	return filepath.Join(s.storageRoot, "files")
}

func (s *StorageManager) TempDir() string {
	return filepath.Join(s.storageRoot, "tmp")
}

func (s *StorageManager) EnsureStorageDirs() error {
	if err := os.MkdirAll(s.FilesDir(), 0755); err != nil {
		return err
	}
	return os.MkdirAll(s.TempDir(), 0755)
}

func (s *StorageManager) GetFilePath(sha string) string {
	sha = strings.ToLower(strings.TrimSpace(sha))
	if len(sha) < 4 {
		return filepath.Join(s.FilesDir(), sha)
	}
	return filepath.Join(s.FilesDir(), sha[:2], sha[2:4], sha)
}

func (s *StorageManager) GetTempPath(fileID uuid.UUID) string {
	return filepath.Join(s.TempDir(), fileID.String())
}

func (s *StorageManager) FileExists(sha string) bool {
	path := s.GetFilePath(sha)
	fi, err := os.Stat(path)
	return err == nil && !fi.IsDir()
}

func ChecksumSHA256(data []byte) string {
	sum := sha256.Sum256(data)
	return hex.EncodeToString(sum[:])
}

func (s *StorageManager) CreateDownloadTicket(itemID, fileID uuid.UUID) (string, time.Time, error) {
	expiresAt := time.Now().UTC().Add(time.Duration(s.ttlSeconds) * time.Second)
	claims := jwt.MapClaims{
		"sub":     "download",
		"item_id": itemID.String(),
		"file_id": fileID.String(),
		"exp":     expiresAt.Unix(),
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	tokenStr, err := token.SignedString(s.jwtSecret)
	if err != nil {
		return "", time.Time{}, err
	}
	return tokenStr, expiresAt, nil
}

func (s *StorageManager) VerifyDownloadTicket(ticket string, itemID, fileID uuid.UUID) bool {
	token, err := jwt.Parse(ticket, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
		}
		return s.jwtSecret, nil
	})
	if err != nil || !token.Valid {
		return false
	}

	claims, ok := token.Claims.(jwt.MapClaims)
	if !ok {
		return false
	}

	if claims["sub"] != "download" {
		return false
	}
	if claims["item_id"] != itemID.String() {
		return false
	}
	if claims["file_id"] != fileID.String() {
		return false
	}
	return true
}

func (s *StorageManager) SaveUploadStream(fileID uuid.UUID, expectedSHA256 string, expectedSize int64, reader io.Reader) (string, error) {
	if err := s.EnsureStorageDirs(); err != nil {
		return "", err
	}

	tempPath := s.GetTempPath(fileID)
	tempFile, err := os.OpenFile(tempPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
	if err != nil {
		return "", fmt.Errorf("failed to create temp file: %w", err)
	}

	hasher := sha256.New()
	mw := io.MultiWriter(tempFile, hasher)

	// Stream with buffer and size check
	buf := make([]byte, 64*1024)
	var written int64
	var writeErr error

	for {
		nr, er := reader.Read(buf)
		if nr > 0 {
			written += int64(nr)
			if written > s.maxFileSize {
				writeErr = errors.New("File size exceeds MAX_FILE_SIZE")
				break
			}
			if written > expectedSize {
				writeErr = fmt.Errorf("Size mismatch: expected %d, got more", expectedSize)
				break
			}
			nw, ew := mw.Write(buf[0:nr])
			if ew != nil {
				writeErr = ew
				break
			}
			if nr != nw {
				writeErr = io.ErrShortWrite
				break
			}
		}
		if er != nil {
			if er != io.EOF {
				writeErr = er
			}
			break
		}
	}

	tempFile.Close()

	if writeErr != nil {
		_ = os.Remove(tempPath)
		return "", writeErr
	}

	if written != expectedSize {
		_ = os.Remove(tempPath)
		return "", fmt.Errorf("Size mismatch: expected %d, got %d", expectedSize, written)
	}

	computedSHA := hex.EncodeToString(hasher.Sum(nil))
	if !strings.EqualFold(computedSHA, expectedSHA256) {
		_ = os.Remove(tempPath)
		return "", fmt.Errorf("SHA-256 mismatch: expected %s, got %s", strings.ToLower(expectedSHA256), computedSHA)
	}

	targetPath := s.GetFilePath(computedSHA)
	if err := os.MkdirAll(filepath.Dir(targetPath), 0755); err != nil {
		_ = os.Remove(tempPath)
		return "", err
	}

	// Move file (atomic replace if on same volume; fallback to copy/delete if cross-volume)
	if err := os.Rename(tempPath, targetPath); err != nil {
		// Fallback copy
		if cpErr := copyFile(tempPath, targetPath); cpErr != nil {
			_ = os.Remove(tempPath)
			return "", cpErr
		}
		_ = os.Remove(tempPath)
	}

	return targetPath, nil
}

func copyFile(src, dst string) error {
	in, err := os.Open(src)
	if err != nil {
		return err
	}
	defer in.Close()

	out, err := os.Create(dst)
	if err != nil {
		return err
	}
	defer out.Close()

	_, err = io.Copy(out, in)
	return err
}

func (s *StorageManager) CleanupEmptyParents(fullPath string) {
	parent := filepath.Dir(fullPath)
	root := s.FilesDir()

	for parent != root && strings.HasPrefix(parent, root) {
		entries, err := os.ReadDir(parent)
		if err != nil || len(entries) > 0 {
			break
		}
		if err := os.Remove(parent); err != nil {
			break
		}
		parent = filepath.Dir(parent)
	}
}

func (s *StorageManager) DeleteFileIfUnreferenced(ctx context.Context, db *sql.DB, sha256 string) (bool, error) {
	sha256 = strings.ToLower(strings.TrimSpace(sha256))
	var count int
	err := db.QueryRowContext(ctx, "SELECT COUNT(*) FROM drop_files WHERE LOWER(sha256) = LOWER($1)", sha256).Scan(&count)
	if err != nil {
		// Try SQLite ? parameter if $1 failed
		err = db.QueryRowContext(ctx, "SELECT COUNT(*) FROM drop_files WHERE LOWER(sha256) = LOWER(?)", sha256).Scan(&count)
	}
	if err != nil {
		return false, err
	}

	if count == 0 {
		path := s.GetFilePath(sha256)
		if fi, err := os.Stat(path); err == nil && !fi.IsDir() {
			_ = os.Remove(path)
			s.CleanupEmptyParents(path)
			return true, nil
		}
	}
	return false, nil
}

func (s *StorageManager) CleanupTempFiles(maxAge time.Duration) int {
	_ = s.EnsureStorageDirs()
	cutoff := time.Now().Add(-maxAge)
	removed := 0

	entries, err := os.ReadDir(s.TempDir())
	if err != nil {
		return 0
	}

	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		info, err := entry.Info()
		if err == nil && info.ModTime().Before(cutoff) {
			path := filepath.Join(s.TempDir(), entry.Name())
			if err := os.Remove(path); err == nil {
				removed++
			}
		}
	}
	return removed
}

func (s *StorageManager) ScanPhysicalFiles() ([]PhysicalFileInfo, error) {
	_ = s.EnsureStorageDirs()
	var results []PhysicalFileInfo
	root := s.FilesDir()

	err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
		if err != nil || info == nil || info.IsDir() {
			return nil
		}
		sha := strings.ToLower(filepath.Base(path))
		relPath, _ := filepath.Rel(root, path)
		results = append(results, PhysicalFileInfo{
			SHA256:   sha,
			Size:     info.Size(),
			Path:     relPath,
			FullPath: path,
		})
		return nil
	})

	return results, err
}

func (s *StorageManager) RemoveOrphanFile(fullPath string) bool {
	fi, err := os.Stat(fullPath)
	if err != nil || fi.IsDir() {
		return false
	}
	if err := os.Remove(fullPath); err == nil {
		s.CleanupEmptyParents(fullPath)
		return true
	}
	return false
}

